core/corehttp/gateway_handler_car.go

```diff
@@ -0,0 +1,59 @@
+package corehttp
+
+import (
+	"context"
+	"net/http"
+
+	blocks "github.com/ipfs/go-block-format"
+	cid "github.com/ipfs/go-cid"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	ipath "github.com/ipfs/interface-go-ipfs-core/path"
+	gocar "github.com/ipld/go-car"
+	selectorparse "github.com/ipld/go-ipld-prime/traversal/selector/parse"
+)
+
+// serveCar returns a CAR stream for specific DAG+selector
+func (i *gatewayHandler) serveCar(w http.ResponseWriter, r *http.Request, rootCid cid.Cid, contentPath ipath.Path) {
+	ctx := r.Context()
+
+	// Set Content-Disposition
+	name := rootCid.String() + ".car"
+	setContentDispositionHeader(w, name, "attachment")
+
+	// Weak Etag W/ because we can't guarantee byte-for-byte identical  responses
+	// (CAR is streamed, blocks arrive from datastore in non-deterministic order)
+	w.Header().Set("Etag", `W/"`+rootCid.String()+`.car"`)
+
+	// Explicit Cache-Control to ensure fresh stream on retry.
+	// CAR stream could be interrupted, and client should be able to resume and get full response, not the truncated one
```

Comment:
Not necessary, but will add `Accept-Ranges: none` just to make it clear we don't support range-requests for CAR stream responses.

I am bummed about caches too.  The underlying problem is this: a valid  CAR stream can be returned as HTTP 200 OK response, and die in the middle of the stream. The end will be missing, but HTTP status code remains 200 (OK).

There is no way to indicate to a browser client (JS code running on some page) that an error occurred (it already received HTTP status OK, and browsers [do not support trailer headers](https://github.com/mdn/browser-compat-data/issues/14703)).  If CDN/HTTP cache stores such truncated CAR response, then other clients will get broken responses too.

I was thinking about a way we could salvage HTTP caching  and here is an idea. We want CAR consumers to do validation anyway, and be able to fetch sub-dags via CAR+selector (https://github.com/ipfs/go-ipfs/issues/8769). 

Such a client will be smart enough to detect missing blocks anyway (during response validation), and request remaining blocks via new request(s). For such client, a truncated CAR response coming from a CDN/cache is not a hard error, it is just part of the daily business.

I feel this is way more robust setup than forcing CAR responses to be deterministic.

Reply:
I was hopping that using accept-ranges none would allow to allow caches because the clients would understand that half of the cache would still need to be fully downloaded but clients would still cache complete requests.

However from what you say that seems to be wrong, so I think this could have been dropped, I don't think it's really a bad thing for it to be present, but it's solving what I had in mind.